cpp/src/parquet/encryption/encryption_internal.cc

```diff
@@ -438,56 +460,105 @@ std::shared_ptr<AesDecryptor> AesDecryptor::Make(
-  int len;
-  int plaintext_len;
+int AesDecryptor::PlaintextLength(int ciphertext_len) const {
+  return impl_->PlaintextLength(ciphertext_len);
+}
 
-  uint8_t tag[kGcmTagLength];
-  memset(tag, 0, kGcmTagLength);
-  uint8_t nonce[kNonceLength];
-  memset(nonce, 0, kNonceLength);
+int AesDecryptor::CiphertextLength(int plaintext_len) const {
+  return impl_->CiphertextLength(plaintext_len);
+}
 
+int AesDecryptor::AesDecryptorImpl::GetCiphertextLength(
+    span<const uint8_t> ciphertext) const {
   if (length_buffer_length_ > 0) {
+    // Note: length_buffer_length_ must be either 0 or kBufferSizeLength
+    if (ciphertext.size() < static_cast<size_t>(kBufferSizeLength)) {
+      std::stringstream ss;
+      ss << "Ciphertext buffer length " << ciphertext.size()
+         << " is insufficient to read the ciphertext length."
+         << " At least " << kBufferSizeLength << " bytes are required.";
+      throw ParquetException(ss.str());
+    }
+
     // Extract ciphertext length
     int written_ciphertext_len = ((ciphertext[3] & 0xff) << 24) |
                                  ((ciphertext[2] & 0xff) << 16) |
                                  ((ciphertext[1] & 0xff) << 8) | ((ciphertext[0] & 0xff));
 
-    if (ciphertext_len > 0 &&
-        ciphertext_len != (written_ciphertext_len + length_buffer_length_)) {
-      throw ParquetException("Wrong ciphertext length");
+    if (written_ciphertext_len < 0) {
+      std::stringstream ss;
+      ss << "Negative ciphertext length " << written_ciphertext_len;
+      throw ParquetException(ss.str());
+    } else if (ciphertext.size() <
+               static_cast<size_t>(written_ciphertext_len) + length_buffer_length_) {
+      std::stringstream ss;
+      ss << "Serialized ciphertext length "
+         << (written_ciphertext_len + length_buffer_length_)
+         << " is greater than the provided ciphertext buffer length "
+         << ciphertext.size();
+      throw ParquetException(ss.str());
     }
-    ciphertext_len = written_ciphertext_len + length_buffer_length_;
+
+    return written_ciphertext_len + length_buffer_length_;
   } else {
-    if (ciphertext_len == 0) {
-      throw ParquetException("Zero ciphertext length");
+    if (ciphertext.size() > static_cast<size_t>(std::numeric_limits<int32_t>::max())) {
+      std::stringstream ss;
+      ss << "Ciphertext buffer length " << ciphertext.size() << " overflows int32";
+      throw ParquetException(ss.str());
     }
+    return static_cast<int>(ciphertext.size());
+  }
+}
+
+int AesDecryptor::AesDecryptorImpl::GcmDecrypt(span<const uint8_t> ciphertext,
+                                               span<const uint8_t> key,
+                                               span<const uint8_t> aad,
+                                               span<uint8_t> plaintext) {
+  int len;
+  int plaintext_len;
+
+  uint8_t tag[kGcmTagLength];
+  memset(tag, 0, kGcmTagLength);
```

Comment:
Or more idiomatically:
```c++
  std::array<uint8_t, kGcmTagLength> tag{};
```

Reply:
I've changed this as well as the other uses of C style arrays in https://github.com/apache/arrow/pull/43195/commits/ff31c7b23484ef8194ec7e098303d44b92e5b8da